### PR TITLE
feat(typegen): retain intersection types

### PR DESCRIPTION
**Pull request by janpaepke**
This PR is based on the discussion here:
https://github.com/mmkal/slonik-tools/issues/386

### CODE REVIEW

**Review by mmkal — COMMENTED** [review]
This is **awesome**! I think we’re gonna immediately start using the to get rid of loads of bogus nulls. Will take a closer look at the docs and details over the next day or two. Let me know if you’re still working on this - one thing that occurs to me is it’d be good to have a slightly more high level test as well. i.e. no templated params, so it’s easy to see at a glance how this feature works. I’d be happy to push this to your branch if you’re not actively reshuffling stuff - let me know.

**Comment by mmkal** [comment]
> So this works surprisingly well. So much so, that I don't fully trust I didn't overlook something. I would ask you to test it please.
> 
> Additionally I restructured the readme quite a bit.
> 
> * I removed the redundancy of the config and cli param definitions by combining them into a table.
> * added docs for this feature.
> 
> Looking through the readme now, I feel like everything that's in "Advanced Configuration" should maybe go into a separate file. What do you think?

I like the readme updates - I am generally not a fan of splitting readme files unless there’s a full docusaurus site because it loses cmd-f without an adequate replacement. Also just to make the diff cleaner, what do you think about pushing in the unrelated readme changes in first so it’s easy to look back at this PR and see what was added?

**Comment by janpaepke** [comment]
Hey Misha,
sure, I apologise for falling into my bad habbit of "while I'm at this, I might as well do that".
You're right, it should have been a separate PR, even though the changes in the readme are in distinct sections, there is no overlap. Look [here](https://github.com/mmkal/slonik-tools/pull/390) for the config only changes and [here](https://github.com/mmkal/slonik-tools/blob/be937325b7b3ebe4b92ba02ded94a2f0d959e7b5/packages/typegen/readme.md) to preview the new readme.

The docs for the new feature are [here](https://github.com/mmkal/slonik-tools/blob/ad3c7dc05072d591dbb58cd47ba7684bc8aa655b/packages/typegen/readme.md#enhancing-return-types).

Re tests: sure, have a go! Don't worry, I can keep my hands busy elsewhere. ;)

**Comment by janpaepke** [comment]
Damn it, I found a caveat.
Intersecting the types will drop the comments.
So basically you lose the info what column the property comes from, which typegen graciously tells you.

Off the top of my head I don't have a solution. Do you?

File: packages/typegen/test/inline-tag-modification.test.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    await checkModification('{col: string} & Other', 'queries.TestTable & Other') // we can't tell if the first intersection type is generated, so it will always be overwritten
+  })
+
+  test('high-level', async () => {
+    const syncer = fsSyncer.jestFixture({
+      targetState: {
+        'index.ts': `
+          import {sql} from 'slonik'
+
+          export default sql<{} & {bar: number}>\`select foo, bar from test_table\`
```
[the remark this answers is not in the document]
Comment (mmkal) [reply packages/typegen/test/inline-tag-modification.test.ts:84]: Nope haha- did in a hurry, will change to string

File: packages/typegen/readme.md [hunk 2]
```diff
@@ -333,6 +333,65 @@ module.exports.default = {
...
+```typescript
+type ResultingType = {
+  string_col: string | null,
+  json_col: string[]
+}
+```
+
+**On subsequent runs typegen will only update the first intersection type and leave all following intersections untouched**.
+
+This also means you can make the column `string_col` non-nullable by intersecting it with `{ string_col: string }`.
```
Comment (mmkal) [inline packages/typegen/readme.md:389]: I had a thought re the dumb mistake I made that you caught - there's an interesting trick people could do and I can't decide if it's an antipattern or not - you could make a column non-nullable by intersecting with `{ string_col: {} }` instead - then typegen would take care of saying whether it's `string` or `number` or `boolean` - and all the `{}` does, effectively, is exclude `null`. But it's kind of weird and non-obvious, and the resultant type shows in vscode as `string & {}` - which acts like `string` but looks confusing. So I don't know if we should "recommend" it. Curious your thoughts though.

File: packages/typegen/src/write/inline.ts [hunk 3]
```diff
@@ -1,15 +1,19 @@
...
 import {tsPrettify} from './prettify'
-import type * as ts from 'typescript'
-import * as path from 'path'
 import {queryInterfaces} from './typescript'
 import {WriteFile} from '.'
 
 // todo: pg-protocol parseError adds all the actually useful information
 // to fields which don't show up in error messages. make a library which patches it to include relevant info.
 
+const queryNamespace = 'queries' // todo: at some point we might want to make this configurable
```
Comment (mmkal) [inline packages/typegen/src/write/inline.ts:15]: 👍

File: packages/typegen/src/write/inline.ts [hunk 4]
```diff
@@ -69,31 +73,51 @@ export function getFileWriter({getQueriesModulePath = defaultGetQueriesModule, w
...
+        if (!matchingQuery) {
+          return
+        }
+        const typeReference = `${queryNamespace}.${matchingQuery.tag}`
+        if (node.typeArguments && node.typeArguments.length === 1) {
+          // existing type definitions
+          const [typeNode] = node.typeArguments
+          if (ts.isIntersectionTypeNode(typeNode)) {
+            // we want to preserve intersection types
+            const [firstArg] = typeNode.types // We can't be sure the first argument is a generated type, but as the namespace might have been overwritten we're gonna have to assume.
```
Comment (mmkal) [inline packages/typegen/src/write/inline.ts:103]: Minor: I don't think this comment reflects (what I think of) as slonik-typegen's general attitude. Even if it's not a generated type, when I'm a user I _want_ slonik typegen to take control and overwrite. This intersection trick and the ignore feature coming up are escape hatches because slonik-typegen, slonik, and postgres itself are imperfect. So I'd suggested removing or making the comment more "aggressive" to show that we're doing it on purpose in case a future person thinks it's a mistake:
```suggestion
            // Always overwrite the first type in the intersection, leave the second alone.
            const [firstArg] = typeNode.types
```

If we make the `queries` namespace configurable in future we could consider making this smarter and less aggressive but until then this feels more honest to me.

Comment (janpaepke) [reply packages/typegen/src/write/inline.ts:103]: I Agree with your point but not with your change.

The first comment is meant to signal to the reader what happens in a specific scenario/routine.
I often do this at the beginning of `if` clauses to signal which case is supposed to be handled here.
Compare: https://github.com/mmkal/slonik-tools/pull/389/files#diff-971380bf36250a35a251a205f99aed60a80f9b5153deacde49eca03408d87975R78

The second comment should be made more "aggressive", as you put it, but it doesn't only leave the second alone, but all subsequent intersections.

File: packages/typegen/readme.md (hunk 2 above)
Comment (janpaepke) [reply packages/typegen/readme.md:389]: Yeah, I share your doubts here. It is very unusual and what I really like about the Intersection solution is that it is very intuitive and readable to users who don't know typegen at all (but have basic understanding of how intersections work). If I saw this code somewhere and something was intersected with `{ col: string }`, I would understand the intention. Where as with `{ string_col: {} }` I would be like "Why is he/she doing that?".
This goes doubly if I don't actually see the intersection but the resulting type being `string & {}` (they might be in different places). I would wonder what this means and follow it to the source.

So in short my recommendation would be to not document, let alone recommend it.

File: packages/typegen/readme.md [hunk 5]
```diff
@@ -333,6 +333,65 @@ module.exports.default = {
 }
 ```
 
+## Enhancing Return Types
+
+Typgen is designed to output types only to the degree it's certain they are correct.
```
Comment (janpaepke) [inline packages/typegen/readme.md:338]:
```suggestion
Typegen is designed to output types only to the degree it's certain they are correct.  
```

haha fail!
And you didn't catch it 😆

File: packages/typegen/src/write/inline.ts (hunk 4 above)
Comment (janpaepke) [reply packages/typegen/src/write/inline.ts:103]:
```suggestion
            // preserve intersection types
            const [firstArg] = typeNode.types // Always overwrite the first type in the intersection, leave all subsequent second alone.
```

Comment (mmkal) [reply packages/typegen/src/write/inline.ts:103]: I’m good w that!

File: packages/typegen/readme.md (hunk 2 above)
Comment (mmkal) [reply packages/typegen/readme.md:389]: Ok - let’s not do it then. Interestingly the typescript team don’t want to introduce a non nullable unknown/any (and it seems I tried to make the case that they should five years ago! https://github.com/microsoft/TypeScript/issues/7648#issuecomment-354036449). If that ever gets picked up then whatever that type were called would be a clearer way to say “typegen can do the hard work, I’ll just say it’s not null”. Anyway no change needed here.

**Comment by janpaepke** [comment]
Sooo. Is this good to go?

**Review by mmkal — APPROVED** [review]
🎉
